README
= Wrangler

== NOTE/DISCLAIMER
This gem is almost completely inspired by/ripped off the exception_notification
plugin/gem, but had to hack too much to get things to work with delayed_job
that I just decided to start from scratch. You'll see that much has been
borrowed however, so I owe a huge debt to the originals (exception_notification
[http://github.com/rails/exception_notification] and
super_exception_notification [http://github.com/pboling/exception_notification])
to help me recreate the Rails hacking.

If you don't really care about using delayed_job for your emailing, consider
going back to the originals, as they're likely better maintained.... ;)

== Overview:
A gem for handling exceptions in a Rails application/environment. 

Some highlights:
* Allows for rendering error pages and sending email notifications in case of
errors.
* Allows for lots of configuring of which pages to show, whether to email or not.
* Allows for asynchronous emailing through delayed_job if available, but works
fine even if delayed_job not installed (email will be synchronous however)
* Will honor filter_parameters set in the controller including the Wrangler
module (except if the parameters are in the URL for GETs) when logging and
emailing application state at the time of the exception
* Allows email notification on exceptions thrown totally outside Controller
context, e.g. in a script run (with rails environment) in a cronjob or
delayed_job
* See TODO for things that are explicitly not included yet, but that could/
should be added

== Quickstart
=== Bare Minimum
There are a lot of defaults set, so getting started should be pretty easy. In
fact, if you don't want any email notifications, you basically just need to
include the Wrangler module in your controller:
    # application_controller.rb
    class ApplicationController < ActionController::Base
      include Wrangler

      ...
    end

=== Enabling email notifications
Email notifications are configured to be sent with the default configuration,
but you'll need to specify from and to addresses for emails to actually be sent.
So that brings us to the configuration of Wrangler. Recommended: just create a
new initializer file (create a wrangler.rb (or whatever you want to name the file)
in  RAILS_ROOT/config/initializers/wrangler.rb). In it, add the following:

  Wrangler::ExceptionNotifier.configure do |notifier_config|
    notifier_config.merge! :from_address => 'notifer@example.com',
                           :recipient_addresses => ['ops-team@example.com']
  end

And, if you haven't already configured ActionMailer to send emails, you'll need
to do that (e.g. setting ActionMailer::Base.smtp_settings), and even after you
have, you may want to change some settings in order to send from a different
account from the one you may use to email your users (e.g. change the :user_name
for smtp_settings):

  Wrangler::ExceptionNotifier.smtp_settings.merge! :user_name => 'notifier@mydomain.com'

(Recommend just putting that in the same wrangler.rb initializer you created above)

For more info on smtp_settings, see ActionMailer (http://am.rubyonrails.org/)

== Configuration
There are two different classes that receive configuration, ExceptionHandler and
ExceptionNotifier.

ExceptionHandler stores configurations about what to do about exceptions (whether to
handle them, email them, which error templates to render for each exception...).

ExceptionNotifier handles the sending of emails when exceptions occur, so
stores configurations about where to send the emails.

You override defaults on each using the same syntax (as seen above) by calling the
configure() method on the class and using the config hash that is yielded to
set your configurations. See the method documentation for the configure() methods
themselves, but here's the basic idea:

     Wrangler::ExceptionHandler.configure do |handler_config|
       handler_config[:key1] = value1
       handler_config[:key2] = value2
       handler_config[:key_for_a_hash].merge! :subkey => value
       handler_config[:key_for_an_array] << another_value
     end
    
     OR
    
     Wrangler::ExceptionHandler.configure do |handler_config|
       handler_config.merge! :key1 => value1,
                             :key2 => value2,
       handler_config[:key_for_a_hash].merge! :subkey => value
       handler_config[:key_for_an_array] << another_value
     end

(same with Wrangler::ExceptionNotifier, except different classname)

Most configurations are single values (e.g. nums or strings),
but some are hashes or arrays. You can either overwrite the hashes/arrays, or
selectively delete, or just append. Recommend just appending to the defaults
in most cases, but if you know what you're doing, you can do whatever you like!

Here is the full set of configuration values for both classes, as well as their
default values (pasted in from the classes, so you can check the code directly
to make sure you've got the latest! :) ):

   ####################
   # ExceptionHandler:
   ####################

      :app_name => '',
      :handle_local_errors => false,
      :handle_public_errors => true,
      # send email for local reqeusts. ignored if :handle_local_errors false
      :notify_on_local_error => false,
      # send email for public requests. ignored if :handle_public_errors false
      :notify_on_public_error => true,
      # send email for exceptions caught outside of a controller context
      :notify_on_background_error => true,
      # configure whether to send emails synchronously or asynchronously
      # using delayed_job (these can be true even if delayed job is not
      # installed, in which case, will just send emails synchronously anyway)
      :delayed_job_for_controller_errors => false,
      :delayed_job_for_non_controller_errors => false,
      # mappings from exception classes to http status codes (see above)
      # add/remove from this list as desired in environment configuration
      :error_class_status_codes => Wrangler::codes_for_exception_classes,
      # explicitly indicate which exceptions to send email notifications for
      :notify_exception_classes => %w(),
      # indicate which http status codes should result in email notification
      :notify_status_codes => %w( 405 500 503 ),
      # where to look for app-specific error page templates (ones you create
      # yourself, for example...there are some defaults in this gem you can
      # use as well...and that are configured already by default)
      :error_template_dir => File.join(RAILS_ROOT, 'app', 'views', 'error'),
      # excplicit mappings from exception class to arbitrary error page
      # templates, different set for html and js responses (Wrangler determines
      # which to use automatically, so you can have an entry in both
      # hashes for the same error class)
      :error_class_html_templates => {},
      :error_class_js_templates => {},
      # you can specify a fallback failsafe error template to render if
      # no appropriate template is found in the usual places (you shouldn't
      # rely on this, and error messages will be logged if this template is
      # used). note: there's an even more failsafe template included in the
      # gem (absolute_last_resort...) below, but DON'T CHANGE IT!!!
      :default_error_template => '',
      # these filter out any HTTP params that are undesired
      :request_env_to_skip => [ /^rack\./,
                                "action_controller.rescue.request",
                                "action_controller.rescue.response" ],
      # mapping from exception classes to templates (if desired), express
      # in absolute paths. use wildcards like on cmd line (glob-like), NOT
      # regexp-style

      # just DON'T change this! this is the error template of last resort!
      :absolute_last_resort_default_error_template =>
        File.join(WRANGLER_ROOT,'rails','app','views','wrangler','500.html')

   #####################
   # ExceptionNotifier:
   #####################

      # who the emails will be coming from. if nil or missing or empty string,
      # effectively disables email notification
      :from_address => '',
      # array of addresses that the emails will be sent to. if nil or missing
      # or empty array, effectively disables email notification.
      :recipient_addresses => [],
      # what will show up at the beginning of the subject line for each email
      # sent note: will be preceded by "[<app_name (if any)>...", where app_name
      # is the :app_name config value from ExceptionHandler (or explicit
      # proc_name given to notify_on_error() method)
      :subject_prefix => "#{(defined?(Rails) ? Rails.env : RAILS_ENV).capitalize} ERROR",
      # can use this to define app-specific mail views using the same data
      # made available in exception_notification()
      :mailer_template_root => File.join(WRANGLER_ROOT, 'views')

== Search algorithm for error templates (given an exception and a status_code):
When trying to find an appropriate error page template to render, Wrangler
goes through several different attempts to locate an appropriate template,

CHANGELOG
HEAD
----

0.1.26
----
* don't try to use logger accessor if logger is nil (default to puts in that case)

0.1.25
----
* adding some error handling in the exception handler where we send email notifications in case the mail-sending itself fails, don't want exceptions to bubble up.
* if error occurs in email sending, will attempt to enqueue with delayed_job (if delayed_job appears to be present)

0.1.24
----
* fixing duplicate message in notification email for handle_error calls

0.1.23
----
* made the css class for error messages more specific to wrangler so there are no conflicts with css classes a website is already using

0.1.22
----
* fixing broken logic to look up error page templates by exception superclass
* fixing status code and error template lookup by exception class when cache_classes is false
** NOTE the change in specifying the config for exception lookups to use the exception class NAME, not the class constant (e.g. "StandardError" and no longer StandardError)

0.1.21
----
* adding ability to call some arbitrary code to fetch arbitrary data (defined by the client code) to insert into notification emails. see :call_for_supplementary_info option

0.1.20
----
* make wrangler a little less rails-specific. won't be able to send notifications, but can at least include Wrangler....

0.1.19
----
* default the error string (becomes part of email subject content) to first error in error_messages if there is no exception object

0.1.18
----
* fixing a bug that arises when code that either isn't an ActionController::Base that includes Wrangler or doesn't include Wrangler module at all passes in a request object to handle_error or handle_exception...because the request_data_from_request method was only available to controllers that included Wrangler. moved the relevant methods out into plain old Wrangler module.

0.1.17
----
* fixing bug in notification template that didn't properly display the exception class name.

0.1.16
----
* adding feature to allow blocking notification-sending when http params match certain patterns (e.g. for ignoring browsers you don't like). note that logs can still be logged to error log, just no notifications.

0.1.15
----
* some comments and notification grammar changes
* allowing notify_on_error to accept an optional message that gets logged on error
* fixing +handle_exception+ so that it logs request contents even if exception is nil

0.1.14
----
* adding method to allow sending notifications on arbitrary error condition detection with or without an actual exception

0.1.13
------
* bug fix to prevent to_str calls on exceptions to be called when they shouldn't (must've been an order of operations error or something :( )

0.1.12
------
* added an exception class for http status 403 - forbidden.

0.1.11
------
* no longer throw exceptions within handle_exception if the exception raised does not respond_to?(:to_str) (e.g. instances of Class).

0.1.7-10
--------
* various attempts to prevent automatic installation of the latest versions of actionpack and activesupport whenever the wrangler gem is installed. giving up at this point, but 0.1.10 will fix a terrible bug introduced in the process of these attempts.

0.1.5
-----
* changed wrangler to log errors before sending notifications in case notification sending somehow results in new exception
* tried a little harder to ensure that exception messages (instead of just object id) is included in logs and exception notifications

0.1.4
-----
* (attempted) bug fix to get the prior version's feature to work properly (to actually override the parent class's settings)

0.1.3
-----
* added a means for setting smtp settings without changing the parent class' settings by overriding the smtp_settings getter method

2009-11-13
----------
* got all of the bugs identified so far (i think) and got a full draft of the README file
* renamed the gem because of name conflict with an existing gem named juggler(!)

2009-11-12
----------
* hunting and killing TODOs

2009-11-11
----------
* First draft of the gem
* Got most things working but a lot of outstanding TODOs



lib/wrangler/exception_handler.rb
module Wrangler

  # A utility method that should only be used internal to wrangler. don't call
  # this; it should only be called once by the Config class and you can get/set
  # it there. returns a mapping from exception classes to http status codes
  #-----------------------------------------------------------------------------
  def self.codes_for_exception_classes
    classes = {
      # These are standard errors in rails / ruby
      NameError.name =>      "503",
      TypeError.name =>      "503",
      RuntimeError.name =>   "500",
      ArgumentError.name =>  "500",
      # the default mapping for an unrecognized exception class
      :default => "500"
    }

    # from exception_notification gem:
    # Highly dependent on the verison of rails, so we're very protective about these'
    classes.merge!({ ActionView::TemplateError.name => "500"})             if defined?(ActionView)       && ActionView.const_defined?(:TemplateError)
    classes.merge!({ ActiveRecord::RecordNotFound.name => "400" })         if defined?(ActiveRecord)     && ActiveRecord.const_defined?(:RecordNotFound)
    classes.merge!({ ActiveResource::ResourceNotFound.name => "404" })     if defined?(ActiveResource)   && ActiveResource.const_defined?(:ResourceNotFound)

    # from exception_notification gem:
    if defined?(ActionController)
      classes.merge!({ ActionController::UnknownController.name => "404" })          if ActionController.const_defined?(:UnknownController)
      classes.merge!({ ActionController::MissingTemplate.name => "404" })            if ActionController.const_defined?(:MissingTemplate)
      classes.merge!({ ActionController::MethodNotAllowed.name => "405" })           if ActionController.const_defined?(:MethodNotAllowed)
      classes.merge!({ ActionController::UnknownAction.name => "501" })              if ActionController.const_defined?(:UnknownAction)
      classes.merge!({ ActionController::RoutingError.name => "404" })               if ActionController.const_defined?(:RoutingError)
      classes.merge!({ ActionController::InvalidAuthenticityToken.name => "405" })   if ActionController.const_defined?(:InvalidAuthenticityToken)
    end

    return classes
  end

  # class that holds configuration for the exception handling logic. may also
  # include a helper method or two, but the main interaction with
  # +ExceptionHandler+ is setting and getting config, e.g.
  #
  #   Wrangler::ExceptionHandler.configure do |handler_config|
  #     handler_config.merge! :key => value
  #   end
  #-----------------------------------------------------------------------------
  class ExceptionHandler

    # the default configuration
    @@config ||= {
      :app_name => '',
      :handle_local_errors => false,
      :handle_public_errors => true,
      # send email for local reqeusts. ignored if :handle_local_errors false
      :notify_on_local_error => false,
      # send email for public requests. ignored if :handle_public_errors false
      :notify_on_public_error => true,
      # send email for exceptions caught outside of a controller context
      :notify_on_background_error => true,
      # configure regular expressions for specific HTTP headers, which if they
      # match will NOT send notification emails (e.g. for filtering out crazy
      # browsers).
      # array of hashes, each with one key (http header string) and a value
      # (regexp to match in http header string). this allows for more than one
      # regexp to be specified for the same header if that's more convenient.
      # any one pattern match will block notification.
      # note that this looks in all request headers, including query params
      # (e.g. form fields, url query params)
      :block_notify_on_request_headers => [],
      # configure whether to send emails synchronously or asynchronously
      # using delayed_job (these can be true even if delayed job is not
      # installed, in which case, will just send emails synchronously anyway)
      :delayed_job_for_controller_errors => true,
      :delayed_job_for_non_controller_errors => true,
      # mappings from exception classes to http status codes (see above)
      # add/remove from this list as desired in environment configuration
      # note that the keys are String names of the classes, not the class
      # constants themselves
      # (e.g. "StandardError" => "500",    and _NOT_ StandardError => "500")
      :error_class_status_codes => Wrangler::codes_for_exception_classes,
      # explicitly indicate which exceptions to send email notifications for
      :notify_exception_classes => %w(),
      # indicate which http status codes should result in email notification
      :notify_status_codes => %w( 405 500 503 ),
      # where to look for app-specific error page templates (ones you create
      # yourself, for example...there are some defaults in this gem you can
      # use as well...and that are configured already by default)
      :error_template_dir => (defined?(RAILS_ROOT) ? File.join(RAILS_ROOT, 'app', 'views', 'error') : nil),
      # explicit mappings from exception class to arbitrary error page
      # templates, different set for html and js responses (Wrangler determines
      # which to use automatically, so you can have an entry in both
      # hashes for the same error class)
      # note that the keys are String names of the classes, not the class
      # constants themselves
      # (e.g. "StandardError" => "file",    and _NOT_ StandardError => "file")
      :error_class_html_templates => {},
      :error_class_js_templates => {},
      # you can specify options to pass to the render() call when rendering
      # error page for errors. This is applied globally, so plan accordingly
      # (one good option is to pass in :layout => 'my_layout' so your app
      # layout gets applied to the error templates)
      :render_error_options => {},
      # you can specify a fallback failsafe error template to render if
      # no appropriate template is found in the usual places (you shouldn't
      # rely on this, and error messages will be logged if this template is
      # used). note: there's an even more failsafe template included in the
      # gem (absolute_last_resort...) below, but DON'T CHANGE IT!!!
      :default_error_template => '',
      # these filter out any HTTP params that are undesired
      :request_env_to_skip => [ /^rack\./,
                                "action_controller.rescue.request",
                                "action_controller.rescue.response" ],
      # mapping from exception classes to templates (if desired), express
      # in absolute paths. use wildcards like on cmd line (glob-like), NOT
      # regexp-style

      # just DON'T change this! this is the error template of last resort!
      # if you do change this, you really should have a good reason for it and
      # really know what you're doing. really.
      :absolute_last_resort_default_error_template =>
        File.join(WRANGLER_ROOT,'rails','app','views','wrangler','500.html'),
      # allows for inserting additional data/comments/status/environment into
      # the notification emails. possibilities include fetching the release number,
      # patch info, current state of different services....
      # pass a proc instance that accepts a single
      # argument, the request object (if any).
      # (e.g. +Proc.new { |request| puts "request is: #{request}" }+, but don't
      # do that, that would be redundant, silly and not secure).
      # NOTE: +request+ can be nil (if background process, for example), so
      # implement accordingly
      :call_for_supplementary_info => nil
    }

    # give access to config
    #-----------------------------------------------------------------------------
    def self.config
      @@config
    end


    # Allows for overriding default configuration settings.
    # in your environment.rb or environments/<env name>.rb, use a block that
    # accepts one argument
    # * recommend against naming it 'config' as you will probably be calling it
    #   within the config block in env.rb...):
    # * note that some of the config values are arrays or hashes; you can
    #   overwrite them completely, delete or insert/merge new entries into the
    #   default values as you see fit...but in most cases, recommend AGAINST
    #   overwriting the arrays/hashes completely unless you don't want to
    #   take advantage of lots of out-of-the-box config
    #
    #   Wrangler::ExceptionHandler.configure do |handler_config|
    #     handler_config[:key1] = value1
    #     handler_config[:key2] = value2
    #     handler_config[:key_for_a_hash].merge! :subkey => value
    #     handler_config[:key_for_an_array] << another_value
    #   end
    #
    # OR
    #
    #   Wrangler::ExceptionHandler.configure do |handler_config|
    #     handler_config.merge! :key1 => value1,
    #                           :key2 => value2,
    #     handler_config[:key_for_a_hash].merge! :subkey => value
    #     handler_config[:key_for_an_array] << another_value
    #   end
    #
    # *NOTE*: sure, you can change this configuration on the fly in your app, but
    # we don't recommend it. plus, if you do and you're using delayed_job, there
    # may end up being configuration differences between the rails process and
    # the delayed_job process, resulting in unexpected behavior. so recommend
    # you just modify this in the environment config files...or if you're doing
    # something sneaky, you're on your own.
    #-----------------------------------------------------------------------------
    def self.configure(&block)
      yield @@config
    end


    # translate the exception class to an http status code, using default
    # code (set in config) if the exception class isn't excplicitly mapped
    # to a status code in config
    #---------------------------------------------------------------------------
    def self.status_code_for_exception(exception)
      if exception.respond_to?(:status_code)
        return exception.status_code
      else
        return config[:error_class_status_codes][exception.class.name] ||
               config[:error_class_status_codes][:default]
      end
    end

  end # end ExceptionHandler class

  ##############################################################################
  # actual exception handling code
  ##############################################################################

  # make all of these instance methods also module functions
  module_function

  # execute the code block passed as an argument, and follow notification
  # rules if an exception bubbles out of the block.
  #
  # == arguments
  # [proc_name]  a name for the chunk of code you're running, included in logs
  #              and in the email notifications' subject line. optional, default
  #              is nil (nothing will be displayed).
  # [message]  a message to include in any logs regarding exceptions thrown.
  #            useful to explain what the context of the code was to help
  #            diagnose. optional, default is nil (no message will be displayed
  #            other than the exception's own message).
  # 
  # == return value
  # * if an exception bubbles out of the block, the exception is re-raised to
  #   calling code.
  # * otherwise, returns nil
  #-----------------------------------------------------------------------------
  def notify_on_error(proc_name = nil, message = nil, &block)
    begin
      yield
    rescue => exception
      options = {}
      options.merge! :proc_name => proc_name unless proc_name.nil?
      options.merge! :error_messages => message unless message.nil?
      handle_exception(exception, options)
    end

    return nil
  end


  # publicly available method for explicitly telling wrangler to handle
  # a specific error condition without an actual exception. it's useful if you
  # want to send a notification after detecting an error condition, but don't
  # want to interrupt the stack by raising an exception. if you did catch an
  # exception and want to do somethign similar, just call handle_exception
  # diretly.
  #
  # the error condition will get logged and may result in notification,
  # according to configuration see notify_on_exception?
  #
  # == arguments
  # [error_messages] a message or array of messages (each gets logged on
  #                  separate log call) capturing the error condition that
  #                  occurred. this will get logged AND sent in any
  #                  notifications sent
  #
  # == options
  # also, any of the options accepted by handle_exception
  #-----------------------------------------------------------------------------
  def handle_error(error_messages, options = {})
    options.merge! :error_messages => error_messages
    handle_exception(nil, options)
  end


  # the main exception-handling method. decides whether to notify or not,
  # whether to render an error page or not, and to make it happen.
  #
  # == arguments
  # [exception] the exception that was caught. can be nil, but should
  #             only be nil if notifications should always be sent,
  #             as notification rules are bypassed this case
  #
  # == options
  # [error_messages] any additional message to log and send in notification.
  #                  can also be an array of messages (each gets logged
  #                  separately)
  # [request] the request object (if any) that resulted in the exception
  # [render_errors] boolean indicating if an error page should be rendered
  #                 or not (Rails only). default => false
  # [proc_name] a string representation of the process/app that was running
  #             when the exception was raised. default value is
  #             Wrangler::ExceptionHandler.config[:app_name].
  #-----------------------------------------------------------------------------
  def handle_exception(exception, options = {})
    request = options[:request]
    render_errors = options[:render_errors] || false
    proc_name = options[:proc_name] || config[:app_name]
    error_messages = options[:error_messages] || ['']

    error_messages = [error_messages] unless error_messages.is_a?(Array)

    if exception.respond_to?(:backtrace)
      backtrace = exception.backtrace
    else
      backtrace = caller
    end

    # extract the relevant request data and also filter out any params
    # that should NOT be logged/emailed (passwords etc.)
    request_data = request_data_from_request(request) unless request.nil?

    supplementary_info = nil

    unless config[:call_for_supplementary_info].nil?
      supplementary_info = config[:call_for_supplementary_info].call(request)
      supplementary_info = [supplementary_info] unless supplementary_info.is_a?(Array)
    end

    unless supplementary_info.blank?
      error_messages << "Supplementary info:"
      error_messages += supplementary_info
    end

    if exception.nil?
      exception_classname = nil
      status_code = nil
      log_error error_messages.inspect
      log_error backtrace
      log_error "Request params were:"
      log_error request_data.to_yaml
      error_string = error_messages.shift
    else
      status_code =
        Wrangler::ExceptionHandler.status_code_for_exception(exception)

      log_exception(exception, request_data, status_code, error_messages)

      if exception.is_a?(Class)
        exception_classname = exception.name
      else
        exception_classname = exception.class.name
      end

      if exception.respond_to?(:message)
        error_string = exception.message
      else
        error_string = exception.to_s
      end
    end

    if send_notification?(exception, request, status_code)
      if notify_with_delayed_job?
        # don't pass in request as it contains not-easily-serializable stuff
        log_error "Wrangler sending email notification asynchronously"
        Wrangler::ExceptionNotifier.send_later(:deliver_exception_notification,
                                              exception_classname,
                                              error_string,
                                              error_messages,
                                              proc_name,
                                              backtrace,
                                              supplementary_info,
                                              status_code,
                                              request_data)
      else
        log_error "Wrangler sending email notification synchronously"
        Wrangler::ExceptionNotifier.deliver_exception_notification(exception_classname,
                                         error_string,
                                         error_messages,
                                         proc_name,
                                         backtrace,
                                         supplementary_info,
                                         status_code,
                                         request_data,
                                         request)
      end
    end

    if render_errors
      render_error_template(exception, status_code)
    end

  rescue Exception => unhandled_exception
    # if it looks like a temporary error interacting with SMTP, then enqueue
    # the error using delayed job if possible
    # (testing by name this way in case the exception isn't loaded into
    # environment, which would cause a NameError and be counterproductive...)
    if unhandled_exception.class.name == 'Net::SMTPAuthenticationError' &&
       Wrangler::ExceptionNotifier.respond_to?(:send_later)

      log_error "Wrangler failed to send error notification: #{unhandled_exception.class.name}:"
      log_error "  #{unhandled_exception.to_s}"

      # note: this is specific to an old-ish version of delayed job...should
      # make wrangler compatible with the old and the new...
      log_error "Wrangler attempting to send via delayed job"
        Wrangler::ExceptionNotifier.send_later(:deliver_exception_notification,
                                              exception_classname,
                                              error_string,
                                              error_messages,
                                              proc_name,
                                              backtrace,
                                              supplementary_info,
                                              status_code,
                                              request_data)
    else
      log_error "/!\\ FAILSAFE /!\\ Wrangler encountered an unhandled exception " +
                "while trying to handle an error. The arguments it received " +
                "were:"
      log_error "  exception: #{exception.inspect}"
      log_error "  options: #{options.inspect}"
      log_error "The unhandled error encountered was #{unhandled_exception.class.name}:"
      log_error "  #{unhandled_exception.to_s}"
    end
  end


  # determine if a notificaiton should be sent
  #-----------------------------------------------------------------------------
  def send_notification?(exception = nil, request = nil, status_code = nil)
    send_notification = notify_in_context?

    if !send_notification
      log_error("Will not notify because notify_in_context? returned false")
    end

    if send_notification && !exception.nil?
      send_notification &&= notify_on_exception?(exception, status_code)
      if !send_notification
        log_error("Will not notify because notify_on_exception? returned false")
      end
    end

    if send_notification && !request.nil?
      config[:block_notify_on_request_headers].each do |headers_and_patterns|
        headers_and_patterns.each_pair do |header, regexp|
          # only send notification if the header does NOT match the regexp
          if request.env.include?(header)
            send_notification &&= (regexp !~ request.env[header])
          end
          if request.env['action_controller.request.query_parameters'].include?(header)
            send_notification &&= (regexp !~ request.env['action_controller.request.query_parameters'][header])
          end
          if request.env['action_controller.request.request_parameters'].include?(header)
            send_notification &&= (regexp !~ request.env['action_controller.request.request_parameters'][header])
          end
        end
      end
      if !send_notification
        log_error "Will not notify because :block_notify_on_request_headers was configured to block this request"
      end
    end

    return send_notification
  end


  # determine if the current context (+local?+, +background+) indicates that a
  # notification should be sent. this applies all of the rules around
  # notifications EXCEPT for what the current exception or status code is
  # (see +notify_on_exception?+ for that)
  #-----------------------------------------------------------------------------
  def notify_in_context?
    if self.respond_to?(:local_request?)
      if (local_request? && config[:notify_on_local_error]) ||
          (!local_request? && config[:notify_on_public_error])
        notify = true
      else
        notify = false
      end
    else
      notify = config[:notify_on_background_error]
    end

    return notify
  end


  # determine if the app is configured to notify for the given exception or
  # status code
  #-----------------------------------------------------------------------------
  def notify_on_exception?(exception, status_code = nil)
    # first determine if we're configured to notify given the context of the
    # exception
    notify = notify_in_context?

    # now if config says notify in this case, check if we're configured to
    # notify for this exception or this status code
    return notify &&
      (config[:notify_exception_classes].include?(exception.class) ||
       config[:notify_status_codes].include?(status_code))
  end

  # determine if email should be sent with delayed job or not (delayed job
  # must be installed and config set to use delayed job
  #-----------------------------------------------------------------------------
  def notify_with_delayed_job?
    use_dj = false

    if self.is_a?(ActionController::Base)
      if config[:delayed_job_for_controller_errors] &&
          ExceptionNotifier.respond_to?(:send_later)
        use_dj = true
      else
        use_dj = false
      end
    else
      if config[:delayed_job_for_non_controller_errors] &&
          ExceptionNotifier.respond_to?(:send_later)
        use_dj = true
      else
        use_dj = false
      end
    end

    return use_dj
  end

end
